kedro-telemetry/kedro_telemetry/plugin.py

```diff
@@ -78,10 +81,52 @@ def _get_or_create_uuid() -> str:
         return ""
 
 
+def _get_or_create_project_uuid(pyproject_path: Path) -> str:
+    """
+    Reads a project UUID from a configuration file or generates and saves a new one if not present.
+    """
+    if pyproject_path.exists():
```

Comment:
What if the user runs, say, `python -m spaceflights -p data_processing` (hence a packaged project) so the original `pyproject.toml` is not there, but they happen to have one in the current working directory?

Reply:
In that case, if consent is True and `pyproject.toml` is located in the same directory as `project_metadata. ' project_path` returns, it will be used to store the generated project UUID.

Reply:
Here's what I did:

1. I cloned a random Python project with a `pyproject.toml`, in this case https://github.com/mwaskom/seaborn/
2. While on it, I ran `python -m test_package_kedro --conf-source ../test-package-kedro/dist/conf-test_package_kedro.tar.gz`
3. The `pyproject.toml` file got modified:

```diff
diff --git a/pyproject.toml b/pyproject.toml
index 0a4e497d..72e91917 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,71 +1,34 @@
 [build-system]
-requires = ["flit_core >=3.2,<4"]
+requires = [ "flit_core >=3.2,<4",]
 build-backend = "flit_core.buildapi"
 
 [project]
 name = "seaborn"
 description = "Statistical data visualization"
-authors = [{name = "Michael Waskom", email = "mwaskom@gmail.com"}]
 readme = "README.md"
-license = {file = "LICENSE.md"}
-dynamic = ["version"]
-classifiers = [
-    "Intended Audience :: Science/Research",
-    "Programming Language :: Python :: 3.8",
-    "Programming Language :: Python :: 3.9",
-    "Programming Language :: Python :: 3.10",
-    "Programming Language :: Python :: 3.11",
-    "Programming Language :: Python :: 3.12",
-    "License :: OSI Approved :: BSD License",
-    "Topic :: Scientific/Engineering :: Visualization",
-    "Topic :: Multimedia :: Graphics",
-    "Operating System :: OS Independent",
-    "Framework :: Matplotlib",
-]
...
```

This behavior is introduced by this PR, it doesn't happen with `kedro-telemetry==0.4.0`.

I know it might seem like a special case, but I think writing to this file can have some unexpected consequences.

Some ideas that come to mind:

- We try to be more careful and constrain in which situations we write to the file
- We give up generating new UUIDs for existing projects, and instead we generate a random one on the fly when it's not available (giving the impression that there are more unique projects than there actually are)
  - Maybe this is related to the other comment
- ❓

Reply:
Thank you, @astrojuanlu for sharing this - I've tested the packaging a bit and found out one more problem. The package building fails when we have `project_uuid` under `[project]` section - so we should not store it there. I see the point of modifying other `pyproject.toml` - now we're breaking their formatting.

Based on the above I have two solutions in mind:

1. Store `project_uuid` in a separate custom section and append it to the end of the file, which should not change the formatting. Keep doing it for all the `pyproject.toml` as we do now
2. Store `project_uuid` in a separate custom section and append it to the end of the file. Differentiate whether it is kedro-related `pyproject.toml` or not by checking some keys such as `kedro_init_version` and modify only them.

Reply:
I thought https://pypi.org/p/toml/ had style-preserving write capabilities? (Or maybe not, see prior discussion at https://github.com/kedro-org/kedro/issues/3690#issue-2175592476)

Regardless, `project_uuid` should probably be stored under `[tools.kedro]`, the `[project]` table is only for PEP 621 keys https://peps.python.org/pep-0621/

Reply:
Yes, you're right - the only page one can customise is `[tool]`.

I've ended up with storing it under `[tools.kedro_telemetry]` which is appended to the end of `pyproject.toml`. That way we do not break the format of the original `pyproject.toml` cause we did it for kedro `pyproject.toml` as well. I didn't put it under `[tool.kedro]` since this approach requires `toml.dump` which actually changes formatting. I've tried some other encoders and it didn't help.

Reply:
I've tested the recent update with packaged project as well - and it works well now.

Reply:
Now, it's an expected behaviour—we do not change the formatting but append this to the end of the file. But we can always modify only the kedro-related  `pyproject.toml` but not the custom. The initial thought was that if people opt in for telemetry for packaged projects, they don't mind storing the `project_id` and `.telemetry` files. However, in reality, I think most people will opt out.

Reply:
> I'm still concerned about how packaged projects work but I'm going to open a separate issue about it.

@astrojuanlu are you concerned whether the project will still run or not or whether we collect telemetry correctly for packaged projects?

Reply:
Rather, the way the telemetry collector modifies its environment